logic_tree_tools.py

```diff
@@ -0,0 +1,920 @@
+                ground_motion_logic_tree_nth_highest,
+            )
+        )
+        modified_logic_tree_pair.ground_motion_logic_tree_note += (
+            f"{ground_motion_logic_tree_nth_highest} (nth) h.w.b. > "
+        )
+
+    return modified_logic_tree_pair
+
+
+def check_weight_validity(logic_tree: LogicTree) -> None:
+    """
+    Check that the weights of branches in each branch_set sum to 1.0.
+
+    Parameters
+    ----------
+    logic_tree : LogicTree
+        The logic tree to check.
+
+    Returns
+    -------
+    None
+
+    Raises
+    ------
+    ValueError
+        If the weights of branches in each branch_set do not sum to 1.0.
+    """
+
+    logic_tree = copy.deepcopy(logic_tree)
+
+    if not np.allclose(
+        [
+            np.sum(np.array([branch.weight for branch in branch_set.branches]))
+            for branch_set in logic_tree.branch_sets
+        ],
+        1.0,
+    ):
+        raise ValueError("The weights of branches in each branch_set do not sum to 1.0.")
+
+
+def get_source_branch_parameters_and_values(logic_tree: SourceLogicTree) -> dict:
+    """
+
+    Gets the parameters that define source branches, and the allowed values for each parameter.
+
+    This function's output is a dictionary where keys are source branch_set short names which indicate the
+    tectonic region type and are as follows:
+        CRU : active shallow crust
+        HIK :  Hikurangi–Kermadec subduction interface
+        PUY : Puysegur subduction interface
+        SLAB : subduction intraslab
+
+    The values are lists of lists where the outer list represents the number of parameters needed to define a source
+    branch and the inner list contains the allowed values for each parameter.
+
+    For example, this function produces the following dictionary for the standard NSHM 2022 source logic tree:
+
+    {'PUY': [['dm0.7'], ['bN[0.902, 4.6]'], ['C4.0'], ['s0.28', 's1.0', 's1.72']],
+    'HIK': [['dmTL'], ['bN[0.95, 16.5]', 'bN[1.097, 21.5]', 'bN[1.241, 27.9]'], ['C4.0'], ['s0.42', 's1.0', 's1.58']],
+    'CRU': [['dmgeodetic', 'dmgeologic'], ['tdFalse', 'tdTrue'], ['bN[0.823, 2.7]', 'bN[0.959, 3.4]', 'bN[1.089, 4.6]'], ['C4.2'], ['s0.66', 's1.0', 's1.41']],
+    'SLAB': [['runiform'], ['d1']]}
+
+    Which indicates, for example, that each branch in the CRU branch set has 7 parameters, where the first parameter can
+    have values 'dmgeodetic', 'dmgeologic' and the last parameter can have values 's0.66', 's1.0', 's1.41'.
+
+    Parameters
+    ----------
+    logic_tree : SourceLogicTree
+        The source logic tree from which to extract branch parameters.
+
+    Returns
+    -------
+    unique_values_dict : dict
+        A dictionary where keys are source branch_set short names and values are lists of lists where the outer list
+        represents the number of parameters needed to define the source branch and the inner list contains the
+        allowed values for each parameter.
+    """
+
+    values_dict = {}
```

Comment:
I'd avoid generic names like this and `values_list`, they make it hard to read the code. Make it as specific as possible, without making a mega name and overthinking it (easier said than done to be fair).

Reply:
That's a good point. I've renamed them